consensus/blake3pow/consensus.go

```diff
@@ -639,26 +642,138 @@ func (blake3pow *Blake3pow) Finalize(chain consensus.ChainHeaderReader, header *
+		utxoSetSize = rawdb.ReadUTXOSetSize(chain.Database(), header.ParentHash(nodeCtx))
 	}
 	if multiSet == nil {
-		return nil, fmt.Errorf("Multiset is nil for block %s", header.ParentHash(nodeCtx).String())
+		return nil, 0, fmt.Errorf("Multiset is nil for block %s", header.ParentHash(nodeCtx).String())
+	}
+
+	utxoSetSize += uint64(len(utxosCreate))
+	utxoSetSize -= uint64(len(utxosDelete))
+	if utxoSetSize > params.MaxUTXOSetSize {
+		trimmedUtxos := make([]*types.SpentUtxoEntry, 0)
+		blake3pow.logger.WithFields(log.Fields{
+			"utxoSetSize":    utxoSetSize,
+			"maxUTXOSetSize": params.MaxUTXOSetSize,
+		}).Info("UTXO set size is greater than the maximum allowed UTXO set size. Begin trimming process.")
+		lastTrimmedBlockHeight := rawdb.ReadLastTrimmedBlock(chain.Database(), header.ParentHash(nodeCtx))
+		if lastTrimmedBlockHeight == 0 {
+			lastTrimmedBlockHeight++
+		}
+		for utxoSetSize > params.MaxUTXOSetSize {
+			prevUtxoSetSize := utxoSetSize
+
+			nextBlockToTrim := rawdb.ReadCanonicalHash(chain.Database(), lastTrimmedBlockHeight+1)
+
+			TrimBlock(chain, batch, lastTrimmedBlockHeight+1, nextBlockToTrim, &utxosDelete, &trimmedUtxos, &utxoSetSize, !setRoots, blake3pow.logger) // setRoots is false when we are processing the block
+
+			lastTrimmedBlockHeight++
+			if lastTrimmedBlockHeight == header.NumberU64(nodeCtx) {
+				blake3pow.logger.Fatal("Last trimmed block height is equal to the current block height. This is a bug.")
+			}
+			blake3pow.logger.WithFields(log.Fields{
+				"prevUtxoSetSize":  prevUtxoSetSize,
+				"utxosDeleted":     prevUtxoSetSize - utxoSetSize,
+				"utxoSetSizeAfter": utxoSetSize,
+				"maxUTXOSetSize":   params.MaxUTXOSetSize,
+				"lastTrimmedBlock": lastTrimmedBlockHeight,
+			}).Info("Trimmed block UTXOs")
+		}
+		if !setRoots {
+			rawdb.WriteLastTrimmedBlock(batch, header.Hash(), lastTrimmedBlockHeight)
+			rawdb.WriteTrimmedUTXOs(batch, header.Hash(), trimmedUtxos)
+		}
 	}
 	for _, hash := range utxosCreate {
 		multiSet.Add(hash.Bytes())
 	}
 	for _, hash := range utxosDelete {
 		multiSet.Remove(hash.Bytes())
 	}
-	blake3pow.logger.Infof("Parent hash: %s, header hash: %s, muhash: %s, block height: %d, setroots: %t, UtxosCreated: %d, UtxosDeleted: %d", header.ParentHash(nodeCtx).String(), header.Hash().String(), multiSet.Hash().String(), header.NumberU64(nodeCtx), setRoots, len(utxosCreate), len(utxosDelete))
+	blake3pow.logger.Infof("Parent hash: %s, header hash: %s, muhash: %s, block height: %d, setroots: %t, UtxosCreated: %d, UtxosDeleted: %d, UTXO Set Size: %d", header.ParentHash(nodeCtx).String(), header.Hash().String(), multiSet.Hash().String(), header.NumberU64(nodeCtx), setRoots, len(utxosCreate), len(utxosDelete), utxoSetSize)
+
+	if utxoSetSize < uint64(len(utxosDelete)) {
+		return nil, 0, fmt.Errorf("UTXO set size is less than the number of utxos to delete. This is a bug. UTXO set size: %d, UTXOs to delete: %d", utxoSetSize, len(utxosDelete))
```

Comment:
This should be a panic, not an error. If this situation occurs, the network is in an illegal state, and continuing to operate is dangerous

Reply:
I agree the network is in an invalid state at this point. However @mechanikalk generally tells me not to panic and instead return an error

Reply:
Suppose a peer sends me a block that deletes more UTXOs than exist (intentionally). It *should* be rejected prior to this step, but if it doesn't get rejected previously for whatever reason, we can reject it here.

Reply:
The rule of thumb for errors vs panics is this:
1) if you are already in an illegal/irrecoverable state, panic to avoid further damage (i.e. RPC node continues to serve invalid data)
1) if an error occurs, but you are not in an illegal state yet, return an error, allowing the caller to hopefully recover from the error

This appears to be case 1, right? If this condition evaluates true, our goose is cooked, and continuing to operate just increases the odds of us cooking other peoples' geese